### PR TITLE
AbstractQueryTestCase should run without type less often

### PR DESCRIPTION
**Pull request by jimczi**
AbstractQueryTestCase is instanciated without mapping type one third
of the time. Since most of the tests are disabled when there is no type
this commit modifies the probability to use rarely() instead of a
fixed probability.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: +1

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: If we only have one allowed type on master now, we could also look into making currentTypes a String, not an array. also changing it to "currentType" might be better then.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: If we change currentTypes to be a String constant, we could also use this here.

**Review by cbuescher — APPROVED** [review]
I left a small suggestion, other than that I agree with this change. I would also port the part about the frequency to the 6.x branches, the part about randomTypes might not work because I think we should still test for multiple types there.

File: test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java
Comment (cbuescher) [inline test/framework/src/main/java/org/elasticsearch/test/AbstractQueryTestCase.java]: I just saw this is also tagged with 6.3, I think this part of the change shouldn't be backported because in 6.x we should still test for multiple types (support for 5.x indices)

**Comment by jimczi** [comment]
Thanks for looking @cbuescher . I've changed the pr to always create a type for the main index (_doc) and added a way to create a query shard context that maps to an index with no type. This way we can explicitly test cases where no type is registered. This allows to always test all methods defined in a concrete class and make the no-type test more explicit. I've also removed the v6.x tag since this should only goes to master. Can you take another look ?

**Comment by javanna** [comment]
heya @jimczi this is a good change, thanks! I was wondering, where do we use no types now? There were way too many "assumes" around that before, on the other hand I think that we introduced it in the first place to test that `toQuery` works ok also when no types are registered, or maybe @cbuescher remembers other reasons for that?

**Comment by cbuescher** [comment]
> we introduced it in the first place to test that toQuery works ok also when no types are registered

No, I think that was the main reason. @javanna maybe time to get rid of this case on master altogether, at least from the existing tests I don't see a compelling reason to keep it (except for the special case in GeoBoundingBoxQueryBuilderTests), wdyt? I'm okay with removing it.

@jimczi I took a look at the last update, great to get rid of all the "assumes".

**Comment by javanna** [comment]
mmm, not sure about removing this completely, I think it proved useful in the past to find bugs when calling toQuery and helped increasing code coverage, on the other hand it is annoying that it requires so many exceptions. Maybe we should have a specific testToQuery run against the index with no types? But then we would still have to conditionally run it only for the queries that support it.

**Comment by jimczi** [comment]
> Maybe we should have a specific testToQuery run against the index with no types? But then we would still have to conditionally run it only for the queries that support it.

I think it's reasonable, I prefer to add special cases for the rare queries that need it rather than adding this assumeTrue on every test. I'll update the pr shortly.

**Comment by cbuescher** [comment]
@jimczi just stumbled upon this while going through old PRs on my list. This looks reade after merge conflicts are resolved and would be good to get in I think.

**Comment by jimczi** [comment]
Thanks for the ping @cbuescher . I merged the conflicts and adapted the tests to cope with the new alias fields. I'll merge if the CI finishes without error.

**Comment by jimczi** [comment]
test this please

**Comment by jimczi** [comment]
test this please

**Pull request merged**
